### ISSUE TITLE
calling \GreHyph with an argument

### ISSUE THREAD

**Issue by eroux** [gregorio-project/gregorio#1140]
It would be useful sometimes to be able to reduce the size of the hyphen, but depending on the space between the syllables: having a long hyphen when the space is very big, and a shorter one when the space can be shrinked. I think the best way to do that is by
- calling `\GreHyph` more consistently when inserting an hyphen (sometimes `-` is used)
- calling `\GreHyph` with an argument (ignored by default?) being the width between the two syllables

**Comment by henryso** [comment]
It seems that the `\GreHyph` vs `-` dichotomy was done on purpose (there are comments to that effect).  Before we start changing it, does anyone remember why that was done?

**Comment by eroux** [comment]
I'm affraid I don't...

**Comment by henryso** [comment]
Researching this a bit more, there is an option to make the `\GreHyph` hyphen protrude when at the end of a line by setting `\greseteolhyphen{zero}`, so if we change all `-` to `\GreHyph` and set that `zero` option, hyphens within a line will protrude into the next syllable, which is most likely not a desirable result.  This also means that gregorio (the executable) emitting `\GreHyph` for all hyphens entered in gabc is probably wrong.  How should we proceed here?

**Comment by henryso** [comment]
There are 3 types of hyphens:

- a hyphen within a syllable (entered in gabc)
- a hyphen at the end of a syllable, but not at the end of a line
- a hyphen at the end of a line

Either of the latter two may be "forced" by entering a hyphen at the end of a syllable in gabc.

My proposal:
- A hyphen entered in gabc within a syllable will be `\GreIntraSyllableHyphen`, which will typeset a hyphen, regardless of any other settings
- A hyphen entered in gabc at the end of a syllable will continue to be `\GreForceHyphen` (as it is now), which will force a hyphen to appear at the end of the current syllable.
- A hyphen at the end of a syllable but not at the end of a line will be `\gre@hyphen#1`, which will typeset a hyphen that is not affected by the `\greseteolhyphen` setting.  `#1` is the amount of space between the syllable texts on either side.
- A hyphen at the end of a syllable will be `\gre@eolhyphen`, which will typeset a hyphen and be affected by the `\greseteolhyphen` setting.

Comments?

**Comment by eroux** [comment]
I agree, but I think the best would be to make something a bit more generic, like a `\gre@hyphen` with two arguments, the first being the type (among the 4 you described), and the second the space between the texts of this syllable and of the next one... `\GreForceHyphen`, etc. would call it internally. How does it sound? Also, we must think of a way for users to override the `\gre@hyph` macro in a user-friendly way...

**Comment by henryso** [comment]
What you're suggesting doesn't entirely make sense so me.  The second argument only makes sense for hyphens at the end of syllables which are not at the end of a line.

What is the second argument for a hyphen that is not at the end of a syllable?  0?  The width of a hyphen?

What is the second argument for a hyphen at the end of a line?  Infinity?  0?

Also, `\GreForceHyphen` does not emit a hyphen.  It sets a flag that causes the hyphen determination code to always generate a hyphen (either interline end-of-line, depending on the final position).

**Comment by eroux** [comment]
The second argument would ignored in these cases yes. Ok for `\GreForceHyphen`, let's ignore it then

**Comment by eroux** [comment]
let's finish it in 5.0

**Comment by henryso** [comment]
Is the thing left to do here the override of `\gre@hyph` in a user-friendly way, or is there something else missing?

**Comment by eroux** [comment]
I remember something about being able to control the length of the hyphen to put at the end of a line when it's automatically added in Lua... I'm affraid I can't have a clearer answer without digging in the code (which won't be this week I'm affraid...)
